src/v/storage/segment_deduplication_utils.cc

```diff
@@ -110,17 +111,20 @@ ss::future<model::offset> build_offset_map(
         const auto& seg = iter->get();
         vlog(gclog.debug, "Adding segment to offset map: {}", seg->filename());
         auto read_lock = co_await seg->read_lock();
-        if (seg->is_closed()) {
-            // Stop early if the segment e.g. has been prefix truncated. We'll
-            // make do with the offset map we have so far.
-            vlog(
-              gclog.debug,
-              "Stopping add to offset map, segment closed: {}",
-              seg->filename());
-            break;
-        }
         segment_full_path idx_path = seg->path().to_compacted_index();
+        std::optional<scoped_file_tracker> to_clean;
+        bool segment_closed = false;
         while (true) {
+            if (seg->is_closed()) {
+                // Stop early if the segment e.g. has been prefix truncated.
+                // We'll make do with the offset map we have so far.
+                vlog(
+                  gclog.debug,
+                  "Stopping add to offset map, segment closed: {}",
+                  seg->filename());
+                segment_closed = true;
+                break;
+            }
```

Comment:
nit: moving it here feels like a separate commit

Reply:
Done